### PR TITLE
chore: Make CI tests run a little more quiet

### PR DESCRIPTION
**Pull request by blt**
This commit was born out of a discussion
[here](https://github.com/timberio/vector/pull/7240#discussion_r622519274). This
changes the CI tests to only output the test name when the test fails, replacing
it with a simple dot, and does not capture test output, meaning individual tests
that `println` etc are now more terse. Notably this does not reduce output
entirely -- VRL tests print their name still, there's seemingly some debug
output that is present elsewhere -- but it's a start.

I wasn't clear from the original discussion if we wanted `--quiet` or not, so I
went ahead and included it here. Happy to back that out if folks feel strongly.

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
My preference would be to still see the printlns when the test fails, just not when it passes. I believe just leaving out `--nocapture` but also removing `--quiet` here will do this, yes?

**Comment by blt** [comment]
> My preference would be to still see the printlns when the test fails, just not when it passes. I believe just leaving out `--nocapture` but also removing `--quiet` here will do this, yes?

I believe that `--quiet` implies `--nocapture`, `--format=terse` and something else additional but I haven't fully tracked it through cargo. Perhaps drop the `--quiet` entirely, add `--format=terse`? I think my major concern with our CI is that it's just hard to find which tests have failed in a run since there's _so much_ output. If it's possible to address that some I'd be happy, is what I'm aiming for.

**Comment by jszwedko** [comment]
👍  `--format=terse` seems good to me. I see:

```
running 2 tests
.F
failures:

---- tests::it_doesnt_work stdout ----
printing some output
thread 'tests::it_doesnt_work' panicked at 'assertion failed: `(left == right)`
  left: `5`,
 right: `4`', src/lib.rs:11:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::it_doesnt_work

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed, to rerun pass '--lib'
```

**Comment by blt** [comment]
@jszwedko I like it. That's what we'll do then.

**Review by jszwedko — APPROVED** [review]

**Comment by jszwedko** [comment]
Hmm, it looks like the bench binaries don't like `--format`.

**Comment by jszwedko** [comment]
I'm also noticing the `trace` macro output seems to circumvent this. I think this is still good to go in once we figure out the issue with running the bench tests, but we could follow up on that later.

**Comment by blt** [comment]
> Hmm, it looks like the bench binaries don't like `--format`.

Indeed. Lemme see, do we even have any correctness tests under the bench feature flag? If not -- and I can't spot any in a grep -- we're going to end up compiling them (valuable in itself, but duplicated work in the bench workflow) and then run the binaries with `--test` to no output? 

I am missing some context here on why we enable these flags, I think.

**Comment by blt** [comment]
> I'm also noticing the `tracing` macro output seems to circumvent this. I think this is still good to go in once we figure out the issue with running the bench tests, but we could follow up on that later.

Yep. There's output from the VRL tests that is structured but doesn't get captured either, which we'd want to look into separately as well, probably.

**Comment by jszwedko** [comment]
> Indeed. Lemme see, do we even have any correctness tests under the bench feature flag? If not -- and I can't spot any in a grep -- we're going to end up compiling them (valuable in itself, but duplicated work in the bench workflow) and then run the binaries with --test to no output?

We do have assertions in the benchmarks. The behavior of running `cargo test --benches`, as I remember it, is that it'll execute each benchmark iteration once, including the assertions.

**Comment by blt** [comment]
> > Indeed. Lemme see, do we even have any correctness tests under the bench feature flag? If not -- and I can't spot any in a grep -- we're going to end up compiling them (valuable in itself, but duplicated work in the bench workflow) and then run the binaries with --test to no output?
> 
> We do have assertions in the benchmarks. The behavior of running `cargo test --benches`, as I remember it, is that it'll execute each benchmark iteration once, including the assertions.

Ah ha. If I had my druthers I'd say benchmarks should be solely for benchmarking and unconcerned with correctness, just to keep the two concerns separate if nothing else.

**Comment by jszwedko** [comment]
> > > Indeed. Lemme see, do we even have any correctness tests under the bench feature flag? If not -- and I can't spot any in a grep -- we're going to end up compiling them (valuable in itself, but duplicated work in the bench workflow) and then run the binaries with --test to no output?
> > 
> > 
> > We do have assertions in the benchmarks. The behavior of running `cargo test --benches`, as I remember it, is that it'll execute each benchmark iteration once, including the assertions.
> 
> Ah ha. If I had my druthers I'd say benchmarks should be solely for benchmarking and unconcerned with correctness, just to keep the two concerns separate if nothing else.

That's fair :) We've just run into situations in the past when the benchmarks didn't actually bench what we thought they were benching, without the assertions. But we don't want the assertions to run as part of the benchmark, so we only run them when `debug_assertions` is enabled, i.e. when running as tests.

**Comment by blt** [comment]
> > > > Indeed. Lemme see, do we even have any correctness tests under the bench feature flag? If not -- and I can't spot any in a grep -- we're going to end up compiling them (valuable in itself, but duplicated work in the bench workflow) and then run the binaries with --test to no output?
> > > 
> > > 
> > > We do have assertions in the benchmarks. The behavior of running `cargo test --benches`, as I remember it, is that it'll execute each benchmark iteration once, including the assertions.
> > 
> > 
> > Ah ha. If I had my druthers I'd say benchmarks should be solely for benchmarking and unconcerned with correctness, just to keep the two concerns separate if nothing else.
> 
> That's fair :) We've just run into situations in the past when the benchmarks didn't actually bench what we thought they were benching, without the assertions. But we don't want the assertions to run as part of the benchmark, so we only run them when `debug_assertions` is enabled, i.e. when running as tests.

I see, that does make sense. I think what I had in mind as an anti-pattern was a situation where a benchmark would _also_ serve as correctness test for the system being examined, but what you're describing is a sanity check situation which totally makes sense. 

I wonder if we can have our cake and eat it too here by making this sanity-check phase happen as a part of the bench workflow.

**Review by bruceg — APPROVED** [review]
I think this is a win if you can get the current CI failures to go green.

**Comment by blt** [comment]
@jszwedko @bruceg I went through and did two things. First, I am reasonably confident that `--quiet` does do what we want. It'll give test-test binaries the `--format=terse` flag and not bench-test binaries but we'll also still get test failure output. Second, I split up the tests between test-test and bench-test, so bench tests now get checked in a separate job. Part of the noise that has been troubling me in #7240 is just benchmark test names, I now understand, so that'll be a big improvement for me to have them split out.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -101,7 +101,29 @@ jobs:
       - run: bash scripts/environment/prepare.sh
       - run: echo "::add-matcher::.github/matchers/rust.json"
       - run: make slim-builds
-      - run: make test-components
+      - run: make test
+
+  # Remove this once https://github.com/timberio/vector/issues/3771 is closed.
+  # Then, modify the `cross-linux` job to run `test` instead of `build`.
+  test-benches-linux:
```
Comment (jszwedko) [inline .github/workflows/test.yml:108]: I think we need to add these new jobs to the notify job at the bottom.

**Review by jszwedko — APPROVED** [review]
👍 this looks good to me. I think we may want to undo part of the changes I made in https://github.com/timberio/vector/pull/7287 though which add `test = true` to all of the benches; otherwise they are picked up by `cargo test`. I like separating them out though it does increase CI pressure a bit building a mostly identical vector build twice.

File: .github/workflows/test.yml (hunk 1 above)
Comment (blt) [reply .github/workflows/test.yml:108]: Quite right. I did miss that.

**Comment by blt** [comment]
> 👍 this looks good to me. I think we may want to undo part of the changes I made in #7287 though which add `test = true` to all of the benches; otherwise they are picked up by `cargo test`.

I can back that out. 

>  I like separating them out though it does increase CI pressure a bit building a mostly identical vector build twice.

Indeed. I have wondered if we couldn't do a single build stage and pass the `target/` forward as an artifact, but I need to look at that more. It would only work for a truly identical build, of course.

File: Makefile [hunk 2]
```diff
@@ -278,7 +278,12 @@ cross-image-%:
 
 .PHONY: test
 test: ## Run the unit test suite
-	${MAYBE_ENVIRONMENT_EXEC} cargo test --workspace --no-fail-fast --no-default-features --features ${DEFAULT_FEATURES} ${SCOPE} -- --nocapture
+	${MAYBE_ENVIRONMENT_EXEC} cargo test --quiet --workspace --no-fail-fast --no-default-features --features ${DEFAULT_FEATURES} ${SCOPE}
+
+.PHONY: test-benches
+test-benches: ## Run benchmarks in debug mode to trigger debug_assertions
+test-benches: export DEFAULT_FEATURES:="benches metrics-benches remap-benches"
```
Comment (bruceg) [inline Makefile:285]: Is this the right syntax? I'm reading this as "target `test-benches` depends on files named `export` and `DEFAULT_FEATURES:="benches metrics-benches remap-benches"`"

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -101,7 +101,29 @@ jobs:
       - run: bash scripts/environment/prepare.sh
       - run: echo "::add-matcher::.github/matchers/rust.json"
       - run: make slim-builds
-      - run: make test-components
+      - run: make test
```
Comment (bruceg) [inline .github/workflows/test.yml:104]: Does this cause us to re-run some tests?

File: src/expiring_hash_map.rs [hunk 4]
```diff
@@ -172,7 +172,7 @@ where
     ///                 println!("Expired: {}", val);
     ///                 break;
     ///             }
-    ///             Some(Err(error)) => panic!(format!("Timer error: {:?}", error)),
+    ///             Some(Err(error)) => panic!("Timer error: {:?}", error),
```
Comment (bruceg) [inline src/expiring_hash_map.rs:175]: Is this related?

Comment (blt) [reply src/expiring_hash_map.rs:175]: It's the sole warning that came up in a run reported as a failure.

File: Makefile (hunk 2 above)
Comment (blt) [reply Makefile:285]: That is a good question. I tried to crib the syntax used in `test-components` but am not an expert in GNU make. My understanding is that this is an example of an "export directive".

File: .github/workflows/test.yml (hunk 3 above)
Comment (blt) [reply .github/workflows/test.yml:104]: I'm not sure I follow the question fully. Re-run in other jobs or in this job?

File: Makefile (hunk 2 above)
Comment (jszwedko) [reply Makefile:285]: Indeed, I was also unfamiliar with this syntax, but it appears to be an example of [Target specific variable values](https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html#Target_002dspecific).

File: .github/workflows/test.yml (hunk 3 above)
Comment (bruceg) [reply .github/workflows/test.yml:104]: Never mind, I was missing seeing something in the `Makefile`.

However, this raises the reverse question: Does anything run `make test-components` after this change? AFAICT both `make test-components` and `make test-benches` run the `test` target, but not vice versa.

File: Makefile (hunk 2 above)
Comment (bruceg) [reply Makefile:285]: Ah, gotcha. I had gone back to the GNU make documentation, and this didn't show up in the prerequisites sections. I see it now. Handy.

File: .github/workflows/test.yml (hunk 3 above)
Comment (blt) [reply .github/workflows/test.yml:104]: You know, I had thought there was something but I believe you're correct: it is an orphan after this.

Comment (blt) [reply .github/workflows/test.yml:104]: 24d7ce0 removes this target.
